### ISSUE TITLE
Option (NULL fields) 

### ISSUE THREAD

**Issue by orthecreedence** [exonum/exonum#1440]
I'm working on a proof of concept project using exonum. One thing that's really starting to hold me back is the lack of being able to serialize `Option` in protobufs:

```
// product.proto
message Product {
	string id = 1;
	string name = 3;
	google.protobuf.Timestamp created = 8;
	google.protobuf.Timestamp updated = 9;
	google.protobuf.Timestamp deleted = 10;
	uint64 history_len = 11;
	exonum.Hash history_hash = 12;
}

// models/product.rs
#[derive(Clone, Debug, ProtobufConvert)]
#[exonum(pb = "proto::product::Product", serde_pb_convert)]
pub struct Product {
    pub id: String,
    pub name: String,
    pub created: DateTime<Utc>,
    pub updated: DateTime<Utc>,
    pub deleted: Option<DateTime<Utc>>,
    pub history_len: u64,
    pub history_hash: Hash,
}
```

Outputs:
```
error[E0277]: the trait bound `std::option::Option<DateTime<Utc>>: exonum::proto::ProtobufConvert` is not satisfied
   --> src\block\models\product.rs:135:24
    |
135 | #[derive(Clone, Debug, ProtobufConvert)]
    |                        ^^^^^^^^^^^^^^^ the trait `exonum::proto::ProtobufConvert` is not implemented for `std::option::Option<DateTime<Utc>>`
    |
    = note: required by `exonum::proto::ProtobufConvert::to_pb`
```

I'm finding myself doing things like `if product.deleted == '1970-01-01T00:00:00Z'::parse() { ... }` but also when receiving data from the client, `let name = if name == "" { None } else { Some (name) };`. This is making things overly verbose, but also tricky when trying to decide if a client actually sent data or not. It also means I have to shove bogus data into the database where in most other cases, a `null` would be fine.

I understand this has to do with the design decisions of [protocol buffers](https://github.com/protocolbuffers/protobuf/issues/1606) and them initializing all values to their default.

I'm wondering if there are workarounds for this inside of exonum, or what the most elegant way to handle these things would be, preferably without a bunch of boilerplate/macros.

Thank you!

**Comment by orthecreedence** [comment]
I've been doing a lot more reading on why proto3 made these decisions and it's starting to make sense to me why things are the way they are, and why `Option` would be difficult to shoehorn into serialization with protobufs. I think my main beef is with things like a hypothetical `TxUpdateUser` where you *can* supply something like a `name` or `occupation`, but if you omit the fields then they aren't updated in the object. I suppose this could still be solved by just checking of `name == ""`, the only problem being if `""` is a valid value (which in my cases, it's not, but I could see it being one) then you have an ambiguity between "nothing was passed" and "something was passed but it also happens to be the default." I guess just being verbose and having something like a `bool name_sent = 2;` in the proto file could solve this.

In other words, not sure if Exonum really even needs to handle this case at all. It'd be nice to know if there's a "right" way to handle this, though. I see mention of using `oneof` fields but I'm not sure how this would be implemented/serialized on the rust side of things.

**Comment by aleksuss** [comment]
@orthecreedence thanks for your question. In terms of Exonum the universal solution to handle this case doesn't exist. You can choose a value which would mean a `None` value. As you have mentioned early a value `1970-01-01T00:00:00Z` would be nice to use if a date doesn't set and so on. Basicly, you can implement `ProtobufConvert` for `Option` but it could have undefined behaviour in some cases. E.g. in some protobuf implementation 0 could be serialized as nothing i.e. empty byte array and an empty string can be represented as empty byte array as well. I'd prefer to avoid such ambiguousness.
